### ISSUE TITLE
Crash on Recompiling project scripts

### ISSUE THREAD

**Issue by michaelsakharov** [michaelsakharov/Prowl#7]
**Describe the bug**
The Editor crashes when recompiling project scripts after a Custom component has been added to a GameObject

**To Reproduce**
1. Create .cs MonoBehvaiour script file in your project
2. Add the MonoBehaviour to a gameobject
3. Modify the script again or press Recompile Scripts
4. Crash

**Comment by michaelsakharov** [comment]
If it doesn't already just delete all objects before recompiling, don't worry about bringing them back just yet, we don't have a scene system yet so there's no proper way to reload the world/scene

**Comment by 10xJosh** [comment]
Howdy! I'm having trouble trying to figure out how to add a .cs script to a component. Could you walk me through how to do it? Also, I noticed while I had the editor opened and I'd change a file I'd get an error, is this similar to what you're getting when you recompile?

![2023-11-28 21_34_11-Prowl (Debugging) - Microsoft Visual Studio](https://github.com/michaelsakharov/Prowl/assets/55113224/aa38a58f-cafc-4bf8-b119-452cddca4546)

**Comment by michaelsakharov** [comment]
> Howdy! I'm having trouble trying to figure out how to add a .cs script to a component. Could you walk me through how to do it? Also, I noticed while I had the editor opened and I'd change a file I'd get an error, is this similar to what you're getting when you recompile?
> 
> ![2023-11-28 21_34_11-Prowl (Debugging) - Microsoft Visual Studio](https://private-user-images.githubusercontent.com/55113224/286477913-aa38a58f-cafc-4bf8-b119-452cddca4546.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEyMzI5NzIsIm5iZiI6MTcwMTIzMjY3MiwicGF0aCI6Ii81NTExMzIyNC8yODY0Nzc5MTMtYWEzOGE1OGYtY2FmYy00YmY4LWIxMTktNDUyY2RkY2E0NTQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI5VDA0Mzc1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM3YTQ0NWU0YTc2ZmRmZDIwM2FjZWFkNDkyZDE4Y2Q4M2U2NzM2NGI5NzBmOTQ4NGVkNDVlYzMxZWEzMTE0NTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.rfLSsjpM74LOTIqETEzPICV2Suzj6GMkHULku19qVI0)

Looks like this is a Differant issue

**Comment by 10xJosh** [comment]
> Hey!, You should be able to just add a .cs file from File Explorer (I pushed a commit that implements the ShowInExplorer option in the Asset Browser):
> Untitled.Project.mp4

I was so close to figuring it out! Thanks for the walkthrough!

**Comment by 10xJosh** [comment]
> > Howdy! I'm having trouble trying to figure out how to add a .cs script to a component. Could you walk me through how to do it? Also, I noticed while I had the editor opened and I'd change a file I'd get an error, is this similar to what you're getting when you recompile?
> > ![2023-11-28 21_34_11-Prowl (Debugging) - Microsoft Visual Studio](https://private-user-images.githubusercontent.com/55113224/286477913-aa38a58f-cafc-4bf8-b119-452cddca4546.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEyMzI5NzIsIm5iZiI6MTcwMTIzMjY3MiwicGF0aCI6Ii81NTExMzIyNC8yODY0Nzc5MTMtYWEzOGE1OGYtY2FmYy00YmY4LWIxMTktNDUyY2RkY2E0NTQ2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI5VDA0Mzc1MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM3YTQ0NWU0YTc2ZmRmZDIwM2FjZWFkNDkyZDE4Y2Q4M2U2NzM2NGI5NzBmOTQ4NGVkNDVlYzMxZWEzMTE0NTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.rfLSsjpM74LOTIqETEzPICV2Suzj6GMkHULku19qVI0)
> 
> Looks like this is a Differant issue

Ok ill create an issue for this one

**Comment by michaelsakharov** [comment]
Was working on something else and skimmed over this code, seems like i already have the GameObject reload stuff, so something else is holding onto the Component, as long as that component is in memory (Garbage collector cant dispose of it) then we cant unload the script assembly.

**Comment by 10xJosh** [comment]
So it looks like the error is being caused by the CSharp-Editor.dll and CSharp.dll in the temp folder. Apparently they're being locked by another thread preventing them from being deleted from temp when the project is recompiled.

The code that's causing the issue can be found in the Prowl/Prowl.Editor/EditorApplication.cs Line 143
```
// Delete everything under Temp\Bin
if (Directory.Exists(Path.Combine(Project.TempDirectory, "bin")))
    Directory.Delete(Path.Combine(Project.TempDirectory, "bin"), true);
Directory.CreateDirectory(Path.Combine(Project.TempDirectory, "bin"));
```
Some of the debugging screenshots:

![2023-11-28 23_27_20-Prowl (Debugging) - Microsoft Visual Studio](https://github.com/michaelsakharov/Prowl/assets/55113224/85c83443-8095-4de4-b512-7accaa24733f)

![2023-11-28 23_29_08-Microsoft Visual Studio Debug Console](https://github.com/michaelsakharov/Prowl/assets/55113224/a8981094-7e7f-4013-b9cf-a92bbd3abc82)

**Comment by michaelsakharov** [comment]
Yeah we need to delete the old assemblies so that we can recompile new ones in its place
It works as intended if you don't use the Component.

I don't think threads are involved since we don't really use threads anywhere at the moment, the project is almost entirely single threaded, which needs to change but thats another issue.

I'm pretty sure the problem is that a Reference to the GameObject which has the component attached to it, still exists somewhere in the code even after deleting all GameObjects.

**Comment by 10xJosh** [comment]
> I'm pretty sure the problem is that a Reference to the GameObject which has the component attached to it, still exists somewhere in the code even after deleting all GameObjects.

Yeah looks like that is the case, im gonna have to look into a bit more. Still not sure where the issue is coming from.

**Comment by michaelsakharov** [comment]
The assembly should be getting unloaded, since if you create the script but don't assign the component, you can keep recompiling the project and changing the script.

And the Unload, Assembly deletion, Compile and Load all works fine.
It only breaks once you add a component.,

**Comment by michaelsakharov** [comment]
Ill take a look at this issue today.

**Comment by michaelsakharov** [comment]
https://learn.microsoft.com/en-us/dotnet/standard/assembly/unloadability#troubleshoot-unloadability-issues

**Comment by 10xJosh** [comment]
I tried setting up the debugger and comparing the memory when it was unloaded with and without the component added and the only difference is that theres a ExternalAssemblyLoadContext1 object attached still which we already know is the case.

![2023-12-02 21_47_59-Prowl Editor exe - PID_ 19948 - WinDbg 1 2308 2002 0](https://github.com/michaelsakharov/Prowl/assets/55113224/f93cb0d9-2090-47bf-ad9f-0382fe376181)


I don't know enough about this so I'll have to look into it later on or after you create the serialization you mentioned on discord. 🤷🏾‍♂️

**Comment by 10xJosh** [comment]
Awesome! Good work!
